frontend/server/src/main/java/org/pytorch/serve/wlm/ModelManager.java

```diff
@@ -205,87 +204,139 @@ private ModelArchive createModelArchive(
+                        null);
+        for (String envVar : envp) {
+            String[] parts = envVar.split("=", 2);
+            if (parts.length == 2) {
+                environment.put(parts[0], parts[1]);
+            }
+        }
+        processBuilder.redirectErrorStream(true);
+
+        Process process = processBuilder.start();
+
+        int exitCode = process.waitFor();
+        String line;
+        StringBuilder outputString = new StringBuilder();
+        BufferedReader brdr = new BufferedReader(new InputStreamReader(process.getInputStream()));
+        while ((line = brdr.readLine()) != null) {
+            outputString.append(line + "\n");
+        }
+
+        if (exitCode == 0) {
+            logger.info(
+                    "Created virtual environment for model {}: {}", model.getModelName(), venvPath);
+        } else {
+            logger.error(
+                    "Virtual environment creation for model {} at {} failed:\n{}",
+                    model.getModelName(),
+                    venvPath,
+                    outputString.toString());
+            throw new ModelException(
+                    "Virtual environment creation failed for model " + model.getModelName());
+        }
+    }
+
     private void setupModelDependencies(Model model)
             throws IOException, InterruptedException, ModelException {
         String requirementsFile =
                 model.getModelArchive().getManifest().getModel().getRequirementsFile();
 
         if (configManager.getInstallPyDepPerModel() && requirementsFile != null) {
-            Path requirementsFilePath =
-                    Paths.get(model.getModelDir().getAbsolutePath(), requirementsFile);
-
+            setupModelVenv(model);
             String pythonRuntime = EnvironmentUtils.getPythonRunTime(model);
-
-            File dependencyPath = model.getModelDir();
-            if (Files.isSymbolicLink(dependencyPath.toPath())) {
-                dependencyPath = dependencyPath.getParentFile();
+            if (!isValidVenvPath(pythonRuntime)) {
+                throw new ModelException(
+                        "Invalid python venv runtime path for model "
+                                + model.getModelName()
+                                + ": "
+                                + pythonRuntime);
             }
 
+            Path requirementsFilePath =
+                    Paths.get(model.getModelDir().getAbsolutePath(), requirementsFile);
             List<String> commandParts = new ArrayList<>();
-
             commandParts.add(pythonRuntime);
             commandParts.add("-m");
             commandParts.add("pip");
             commandParts.add("install");
             commandParts.add("-U");
-            commandParts.add("-t");
-            commandParts.add(dependencyPath.getAbsolutePath());
+            commandParts.add("--upgrade-strategy");
+            commandParts.add("only-if-needed");
```

Comment:
I think this is a good change but do you mind just telling me why you needed to make it?

Reply:
In prior versions of pip (i.e pip<10.0) when `pip install -U -r requirements.txt` is used, all the packages listed in `requirements.txt` and their dependencies are upgraded since by default the `--upgrade-strategy` was `eager`. `--upgrade-strategy` applies to the handling of dependencies of the packages specified in `requirements.txt`. In `pip>=10.0`, the default `--upgrade-stragegy` is `only-if-needed`. This change is to explicitly make the `--upgrade-strategy` as `only-if-needed` irrespective of pip version.

From the pip docs: https://pip.pypa.io/en/stable/user_guide/#only-if-needed-recursive-upgrade